pkg/experiment/metrics/observer.go

```diff
@@ -0,0 +1,197 @@
+	"github.com/grafana/pyroscope/pkg/experiment/block"
+	phlaremodel "github.com/grafana/pyroscope/pkg/model"
+)
+
+type SampleObserver struct {
+	exporter          *Exporter
+	tenant            string
+	Recordings        []*Recording
+	recordingTime     int64
+	pyroscopeInstance string
+	logger            log.Logger
+}
+
+type Recording struct {
+	rule  RecordingRule
+	data  map[model.Fingerprint]*Sample
+	state *recordingState
+}
+
+type recordingState struct {
+	fp      *model.Fingerprint
+	matches bool
+	sample  *Sample
+}
+
+func NewSampleObserver(meta *metastorev1.BlockMeta, logger log.Logger) *SampleObserver {
+	recordingTime := int64(ulid.MustParse(meta.Id).Time())
+	pyroscopeInstance := pyroscopeInstanceHash(meta.Shard, meta.CreatedBy)
+	return &SampleObserver{
+		recordingTime:     recordingTime,
+		pyroscopeInstance: pyroscopeInstance,
+		logger:            logger,
+		exporter:          &Exporter{},
+	}
+}
+
+func pyroscopeInstanceHash(shard uint32, createdBy int32) string {
+	buf := make([]byte, 0, 8)
+	buf = append(buf, byte(shard>>24), byte(shard>>16), byte(shard>>8), byte(shard))
+	buf = append(buf, byte(createdBy>>24), byte(createdBy>>16), byte(createdBy>>8), byte(createdBy))
+	return fmt.Sprintf("%x", xxhash.Sum64(buf))
+}
+
+func (o *SampleObserver) Init(tenant string) {
+	o.tenant = tenant
+	recordingRules := recordingRulesFromTenant(o.tenant)
+	o.Recordings = make([]*Recording, len(recordingRules))
+	for i, rule := range recordingRules {
+		o.Recordings[i] = &Recording{
+			rule: *rule,
+			data: make(map[model.Fingerprint]*Sample),
+			state: &recordingState{
+				fp: nil,
+			},
+		}
+	}
+
+}
+
+func (o *SampleObserver) Observe(row block.ProfileEntry) {
+	for _, recording := range o.Recordings {
+		if recording.state.fp == nil || *recording.state.fp != row.Fingerprint {
+			recording.InitState(row.Fingerprint, row.Labels, o.pyroscopeInstance, o.recordingTime)
+		}
+		if !recording.state.matches {
+			continue
+		}
+		recording.state.sample.Value += float64(row.Row.TotalValue())
+	}
+}
+
+func (o *SampleObserver) Flush() error {
+	recs := o.Recordings
+	o.Recordings = nil
+	go func(tenant string, recordings []*Recording) {
+		timeSeries := make([]prompb.TimeSeries, 0)
+		for _, recording := range recordings {
+			for _, sample := range recording.data {
+				ts := prompb.TimeSeries{
+					Labels: make([]prompb.Label, 0, len(sample.Labels)),
+					Samples: []prompb.Sample{
+						{
+							Value:     sample.Value,
+							Timestamp: sample.Timestamp,
+						},
+					},
+				}
+				for _, l := range sample.Labels {
+					ts.Labels = append(ts.Labels, prompb.Label{
+						Name:  l.Name,
+						Value: l.Value,
+					})
+				}
+				timeSeries = append(timeSeries, ts)
+			}
+		}
+
+		if err := o.exporter.Send(tenant, timeSeries); err != nil {
+			level.Error(o.logger).Log("msg", "error sending recording metrics", "err", err)
+		}
+	}(o.tenant, recs)
+	return nil
+}
```

Comment:
I believe we should control concurrency in the `Exporter` implementation. Otherwise, it's implied that the `SampleObserver` knows implementation details of the `Exporter`. Moreover, a method called `Flush` is typically meant to be blocking.